R/plotNhoods.R

```diff
@@ -741,3 +741,109 @@ plotNhoodMA <- function(da.res, alpha=0.05, null.mean=0){
 
   return(ma.p)
 }
+
+
+#' Plot the number of cells in a neighbourhood per sample and condition
+#'
+#' @param x A \code{\linkS4class{Milo}} object with a non-empty \code{nhoodCounts}
+#' slot.
+#' @param subset.nhoods A logical, integer or character vector indicating the rows of \code{nhoodCounts(x)} to use for
+#' plotting.
+#' @param design.df A \code{data.frame} which matches samples to a condition of interest.
+#' The row names should correspond to the samples. You can use the same \code{design.df}
+#' that you already used in the \code{testNhoods} function.
+#' @param condition String specifying the condition of interest Has to be a column in the \code{design}.
+#' @param n_col Number of columns in the output \code{ggplot}.
+#'
+#' @return A \code{ggplot-class} object
+#'
+#' @author
+#' Nick Hirschmüller
+#'
+#' @examples
+#'
+#' require(SingleCellExperiment)
+#' ux.1 <- matrix(rpois(12000, 5), ncol=300)
+#' ux.2 <- matrix(rpois(12000, 4), ncol=300)
+#' ux <- rbind(ux.1, ux.2)
+#' vx <- log2(ux + 1)
+#' pca <- prcomp(t(vx))
+#'
+#' sce <- SingleCellExperiment(assays=list(counts=ux, logcounts=vx),
+#'                             reducedDims=SimpleList(PCA=pca$x))
+#' milo <- Milo(sce)
+#' milo <- buildGraph(milo, k=20, d=10, transposed=TRUE)
+#' milo <- makeNhoods(milo, k=20, d=10, prop=0.3)
+#' milo <- calcNhoodDistance(milo, d=10)
+#'
+#' cond <- sample(c("A","B","C"),300,replace=TRUE)
+#'
+#' meta.df <- data.frame(Condition=cond, Replicate=c(rep("R1", 100), rep("R2", 100), rep("R3", 100)))
+#' meta.df$SampID <- paste(meta.df$Condition, meta.df$Replicate, sep="_")
+#' milo <- countCells(milo, meta.data=meta.df, samples="SampID")
+#'
+#' design.mtx <- data.frame("Condition"=c(rep("A", 3), rep("B", 3), rep("C",3)),
+#'                          "Replicate"=rep(c("R1", "R2", "R3"), 3))
+#' design.mtx$SampID <- paste(design.mtx$Condition, design.mtx$Replicate, sep="_")
+#' rownames(design.mtx) <- design.mtx$SampID
+#'
+#' plotNhoodCounts(x = milo,
+#'                 subset.nhoods = c(1,2),
+#'                 design.df = design.mtx,
+#'                 condition = "Condition")
+#'
+#' @export
+#' @rdname plotNhoodCounts
+#' @importFrom ggplot2 ggplot geom_point stat_summary facet_wrap ylab
+#' @importFrom tidyr pivot_longer
+#' @importFrom tibble rownames_to_column has_rownames
+#' @importFrom dplyr left_join
+plotNhoodCounts <- function(x, subset.nhoods, design.df, condition, n_col=3){
+  if (!is(x, "Milo")) {
+    stop("Unrecognised input type - must be of class Milo")
+  }
+  if (ncol(nhoods(x)) == 1 & nrow(nhoods(x)) == 1) {
+    stop("No neighbourhoods found. Please run makeNhoods() first.")
+  }
+  if (ncol(nhoodCounts(x)) == 1 & nrow(nhoodCounts(x)) == 1) {
+    stop("Neighbourhood counts missing - please run countCells() first")
+  }
+  if (all(subset.nhoods %in% c(TRUE, FALSE))){
```

Comment:
you can just test if this is of type logical with `is(..., "logical")` which is clearer.

Reply:
You also don't need to cast this logical vector to a new integer vector - just check that the length is the same as the number of rows of `nhoodCounts(x)` and give an appropriate error message if it is not.

Reply:
This is also an ideal place to include a unit test on the logical vector length <- note this only applies when `subset.nhoods` is logical. If it is numeric or character then the subsetting on the rows will just work (assuming the elements of `subset.nhoods` are in the rows of `nhoodCounts`.

Reply:
I actually did the conversion on purpose, so the user can specify a logical vector that is shorter than `nhoodCounts(x)`. 
If the logical vector is the same length as `nrow(nhoodCounts(x))`, the function still works without a problem and the time it takes to cast the logical vector to an integer vector is neglectable I thought. If the user specifies a logical vector which for some reason is longer than the number of rows in `nhoodCounts(x)`, it would get picked up by `!all(subset.nhoods %in% rownames(nhoodCounts(x)))` and the user would get an error. However, if you think it makes sense to change this functionality, I will!

Reply:
It doesn't make sense to allow a logical vector that is less than the number of rows in nhoods as there is no way to 100% guarantee that the elements of the logical vector map onto the correct row indices. If I define a logical vector of length 3: `c(TRUE, TRUE, TRUE)` and test if these elements are in the rownames of `nhoodCounts(x)` I will just return a logical vector that is identical to the input vector. If you cast this logical vector with `which(c(TRUE, TRUE, TRUE))` the output is `c(1, 2, 3)` <- that doesn't make sense if I actually want the 12th, 54th and 107th nhoods.

Casting the logical vector is just redundant - it doesn't matter if the overhead is small, it's just a pointless operation and any redundant code creates a) a source of bugs and b) extra maintenance overhead. Force the length of the logical vector to be the same as nrows of nhoodCounts because you will remove a source of bugs.

>  If the user specifies a logical vector which for some reason is longer than the number of rows in nhoodCounts(x), it would get picked up by !all(subset.nhoods %in% rownames(nhoodCounts(x))) and the user would get an error.

Always be explicit and never make assumptions unless you can 100% verify them, e.g. with a unit test.

Reply:
I see your point! Thank you very much for taking the time and explaining it. I have adjusted the code accordingly.